Documentation/ceph-cluster-crd.md

```diff
@@ -1489,3 +1490,40 @@ data:
 
 Note: if you are using self-signed certificates (not known/approved by a proper CA) you must pass `VAULT_SKIP_VERIFY: true`.
 Communications will remain encrypted but the validity of the certificate will not be verified.
+
+##### Ceph CSI configuration
+
+Ceph-CSI is capable of encrypting requested PVs.
+Rook can configure Ceph-CSI to turn on encryption with Key Management System support.
+For this, you need to add new parameter in your CephCluster CR spec:
+
+```yaml
+security:
+  kms:
+  ...
+  ...
+  csiConnectionDetailsCMName: "rook-ceph-csi-vault"
+```
+
+The config map will look like:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
```

Comment:
If I understand correctly, CSI never reads this ConfigMap, only Rook. As mentioned in our chat, I think it makes more sense to include this in the CephCluster CR so it's easier to see the details without having to look in two places.

Reply:
The interesting thing here is that the CSI driver really owns the schema. Rook just needs to pass the settings on to the csi driver after Rook loads the desired state (either from a CM or a CR). There are a few options in my mind:
1. Define new CRD(s) with strongly typed settings
1. Define strongly typed settings in the cluster CR
1. Define weakly typed settings in the cluster CR for more flexibility (e.g. maps of strings)
1. Include the json blob in the cluster CR
1. Load the json from a configmap (basically what's implemented already in this PR)

Options 1-3 are basically different ways for Rook to define the desired state through CRDs. Option 1 is the most appealing from an API perspective, but at the same time may be overkill for this advanced setting. If Rook defines the settings in one of these formats, we basically would be redefining the schema in the format of our choice, then translate it into the json expected by the csi driver. 

Option 4 is rather ugly to include the json blob in the CR and doesn't really gain much over option 5. 

I'm leaning towards option 5 that will keep it simplest in the short term. It's also an advanced feature that may not be used often. We can always move to the CRD(s) later if we decide it's important enough. 

Another reason to stick with raw json is that since CSI owns the format, the json must stay consistent with the CSI version. If rook does any translation of format, we would potentially have to generate different json depending on the csi driver version.

Reply:
Option 1 seems overkill, and it also requires defining a schema in Rook/Kubernetes that CSI already defines via JSON.

These are similar concerns that I had when implementing the `DriveGroups` feature and the JSON blobs that get passed to `ceph-volume`. I think I would cast my vote for option 3, which is what I suggested above. The added value here over configmap I see is that k8s validate that that the definition parses properly as JSON/YAML and that the options are visible in the same place as the cluster CR rather than slightly hidden.

I don't think option 3 would obligate Rook to translate or interpret the JSON in any way. As a counter example, Rook could translate the JSON in option 5 if it comes in via ConfigMap anyway if we really wanted it to. I think the translation point is more about what we officially support (and document) than any requirement.

I do think option 4 or 5 are the next-best, and I don't see a whole lot of difference between them, as the interface is primarily a "JSON blob" that doesn't get validated. If we choose a "JSON blob" option, we could have Rook validate that the JSON parses properly as a convenience validation.

Reply:
Option 3 sticks a bit more with what we already have, for instance, the flag `connectionDetails` is a good example since its content comes straight from a ConfigMap when used with another operator.

At the end, options 3 and 5 are really similar, but I'm leaning toward option 3 to stick with an identical type to `connectionDetails`.

Let's discuss one more time today